cas-server-webapp-support/src/main/java/org/jasig/cas/web/report/SingleSignOnSessionsReportController.java

```diff
@@ -0,0 +1,157 @@
+        AUTHENTICATION_DATE("authentication_date"),
+        TICKET_GRANTING_TICKET("ticket_granting_ticket"),
+        NUMBER_OF_USES("number_of_uses");
+
+        private String attributeKey;
+
+        /**
+         * Instantiates a new Sso session attribute keys.
+         *
+         * @param attributeKey the attribute key
+         */
+        private SsoSessionAttributeKeys(final String attributeKey) {
+            this.attributeKey = attributeKey;
+        }
+
+        @Override
+        public String toString() {
+            return this.attributeKey;
+        }
+    }
+
+    private static final String ROOT_REPORT_ACTIVE_SESSIONS_KEY = "activeSsoSessions";
+
+    private static final String ROOT_REPORT_NA_KEY = "notAvailable";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SingleSignOnSessionsReportController.class);
+
+    private final ObjectMapper jsonMapper = new ObjectMapper();
+
+    private String viewPath = "/WEB-INF/view/jsp/monitoring/viewSsoSessions.jsp";
+
+    @Autowired
+    private CentralAuthenticationService centralAuthenticationService;
+
+    @Value("${sso.sessions.include.tgt:false}")
+    private boolean includeTicketGrantingTicketId;
+
+    /**
+     * Instantiates a new Single sign on sessions report resource.
+     */
+    public SingleSignOnSessionsReportController() {
+        this.jsonMapper.enable(SerializationFeature.INDENT_OUTPUT);
+        this.jsonMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+    }
+
+    /**
+     * Gets sso sessions.
+     *
+     * @return the sso sessions
+     */
+    private Collection<Map<String, Object>> getSsoSessions() {
+        final List<Map<String, Object>> activeSessions = new ArrayList<Map<String, Object>>();
+
+        for(final Ticket ticket : getNonExpiredTicketGrantingTickets()) {
+            final TicketGrantingTicket tgt = (TicketGrantingTicket) ticket;
+
+            final Map<String, Object> sso = new HashMap<>(SsoSessionAttributeKeys.values().length);
+            sso.put(SsoSessionAttributeKeys.AUTHENTICATED_PRINCIPAL.toString(), tgt.getAuthentication().getPrincipal().getId());
+            sso.put(SsoSessionAttributeKeys.AUTHENTICATION_DATE.toString(), tgt.getAuthentication().getAuthenticationDate());
+            sso.put(SsoSessionAttributeKeys.NUMBER_OF_USES.toString(), tgt.getCountOfUses());
+            if (this.includeTicketGrantingTicketId) {
+                sso.put(SsoSessionAttributeKeys.TICKET_GRANTING_TICKET.toString(), tgt.getId());
+            }
+
+            activeSessions.add(Collections.unmodifiableMap(sso));
+        }
+        return Collections.unmodifiableCollection(activeSessions);
+    }
+
+    /**
+     * Gets non expired ticket granting tickets.
+     *
+     * @return the non expired ticket granting tickets
+     */
+    private Collection<Ticket> getNonExpiredTicketGrantingTickets() {
+        return this.centralAuthenticationService.getTickets(new Predicate() {
```

Comment:
Right now, CASImpl is the intermediary between web requests and the ticket registry. So it handles most if not all of the ticket-related ops. Depending on the underlying ticket registry, yes, this might take a bit of time.